### PR TITLE
Homework10

### PR DESCRIPTION
**Pull request by josefschneider**

### CODE REVIEW

File: course5/homework10/linked_list/list.h [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+		--mSize;
+	}
+
+	// TODO what should we return if mSize == 0?
+	T& front() { return mFirstNode->mObj; }
+	const T& front() const { return mFirstNode->mObj; }
+	size_t size() const { return mSize; }
+
+
+//private:
```
Comment (maciekgajewski) [inline course5/homework10/linked_list/list.h:80]: Why not private?

File: course5/homework10/linked_list/list.h [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+		Node(std::shared_ptr<Node> next, const T& copy):
+			mObj(copy),
+			mNextNode(next) { std::cout << "Copy constructed node" << std::endl; }
+
+		Node(std::shared_ptr<Node> next, T&& move):
+			mObj(move),
+			mNextNode(next) { std::cout << "Move constructed node" << std::endl; }
+
+		T mObj;
+		std::shared_ptr<Node> mNextNode;
```
Comment (maciekgajewski) [inline course5/homework10/linked_list/list.h:101]: Why shared? Who else owns it?

File: course5/homework10/linked_list/list.h [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	};
+
+	using Iterator = ListIterator;
+	using Const_iterator = ListIterator;
+
+	Iterator begin() { return Iterator(mFirstNode.get()); }
+	Iterator end() { return Iterator(nullptr); }
+
+	Const_iterator begin() const { return Const_iterator(mFirstNode.get()); }
+	Const_iterator end() const { return Const_iterator(nullptr); }
```
Comment (maciekgajewski) [inline course5/homework10/linked_list/list.h:129]: So you can modify the content trough a const iterator?

File: course5/homework10/string_with_iterators/string.h [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	const char& operator[](const size_t idx) const;
+
+	bool operator==(const String&) const;
+	bool operator!=(const String& rhs) const { return !(*this == rhs); }
+
+	const char* c_str() const { return mSize > 0 ? mContents.get() : ""; }
+	size_t length() const { return mSize > 0 ? mSize - 1 : 0; }
+
+	template <typename T>
+	class StringIterator:
```
Comment (maciekgajewski) [inline course5/homework10/string_with_iterators/string.h:36]: Oh no! String stores data in a continous memory block. This means that you can simply use ```char*``` and ```const char*``` as ```iterator``` and ```const_iterator```

**Review by maciekgajewski — CHANGES_REQUESTED** [review]
- Use unique_ptr instead of shared_ptr in the list. Yu don't really use the refcounting properties of shared_ptr, while you're still paying for them dearly.

File: course5/homework10/string_with_iterators/string.h (hunk 4 above)
Comment (josefschneider) [reply course5/homework10/string_with_iterators/string.h:36]: But would the absence of random_access_iterator_tag not make some operations less efficient, e.g., advance()?

Comment (maciekgajewski) [reply course5/homework10/string_with_iterators/string.h:36]: No; the iterator-property-detecting meta code will recognise pointer as a random-access iterator.

**Pull request closed without merging**